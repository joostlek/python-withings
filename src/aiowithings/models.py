"""Asynchronous Python client for Withings."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from enum import IntEnum, IntFlag, StrEnum
from typing import Any, Self

from aiowithings.util import get_measurement_from_dict, to_enum


class DeviceModel(IntEnum):
    """Enum with device models."""

    UNKNOWN = 0
    WBS01 = 1
    WS30 = 2
    KID_SCALE = 3
    SMART_BODY_ANALYZER = 4
    BODY_PLUS = 5
    BODY_CARDIO = 6
    BODY = 7
    BODY_SCAN = 10
    BODY_PRO = 9
    WBS10 = 11
    WBS11 = 12
    SMART_BABY_MONITOR = 21
    WITHINGS_HOME = 22
    WITHINGS_BLOOD_PRESSURE_MONITOR_V1 = 41
    WITHINGS_BLOOD_PRESSURE_MONITOR_V2 = 42
    WITHINGS_BLOOD_PRESSURE_MONITOR_V3 = 43
    BPM_CORE = 44
    BPM_CONNECT = 45
    BPM_CONNECT_PRO = 46
    PULSE = 51
    ACTIVITE = 52
    ACTIVITE_POP_STEEL = 53
    WITHINGS_GO = 54
    ACITIVTE_STEEL_HR = 55
    ACTIVITE_STEEL_HR_SPORT_EDITION = 59
    PULSE_HR = 58
    MOVE = 90
    MOVE_ECG_1 = 91
    MOVE_ECG_2 = 92
    SCANWATCH = 93
    AURA_DOCK = 60
    AURA_SENSOR = 61
    AURA_SENSOR_V2 = 63
    THERMO = 70
    WUP01 = 100
    IGLUCOSE_GLUCOMETER = 1061
    IOS_STEP_TRACKER_1 = 1051
    IOS_STEP_TRACKER_2 = 1052
    HEALTHKIT_STEP_IPHONE_TRACKER = 1057
    HEALTHKIT_STEP_APPLE_WATCH_TRACKER = 1058
    HEALTHKIT_OTHER_STEP_TRACKER = 1059
    ANDROID_STEP_TRACKER_1 = 1053
    ANDROID_STEP_TRACKER_2 = 1054
    ANDROID_STEP_TRACKER_3 = 1060
    GOOGLE_FIT_TRACKER = 1055
    SAMSUNG_HEALTH_TRACKER = 1056
    HUAWEI_TRACKER = 1062


class DeviceType(StrEnum):
    """Enum representing the device type."""

    UNKNOWN = "unknown"
    SCALE = "Scale"
    BABYPHONE = "Babyphone"
    BLOOD_PRESSURE_MONITOR = "Blood Pressure Monitor"
    ACTIVITY_TRACKER = "Activity Tracker"
    SLEEP_MONITOR = "Sleep Monitor"
    SMART_CONNECTED_THERMOMETER = "Smart Connected Thermometer"
    GATEWAY = "Gateway"
    IGLUCOSE = "iGlucose"
    HEALTHKIT_APPLE = "HealthKit Apple"
    HEALTHKIT_GOOGLE = "HealthKit Google"
    HEALTHKIT_HUAWEI = "HealthKit Huawei"


class DeviceBattery(StrEnum):
    """Enum representing the battery state of the device."""

    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


@dataclass(slots=True)
class Device:
    """Model for a Withings device."""

    device_type: DeviceType
    battery: DeviceBattery
    model: DeviceModel
    first_session_date: datetime | None
    last_session_date: datetime | None
    device_id: str
    hashed_device_id: str

    @classmethod
    def from_api(cls, device: dict[str, Any]) -> Self:
        """Initialize from the API."""
        first_session_date = None
        if device["first_session_date"] is not None:
            first_session_date = datetime.fromtimestamp(
                device["first_session_date"],
                tz=timezone.utc,
            )
        last_session_date = None
        if device["last_session_date"] is not None:
            last_session_date = datetime.fromtimestamp(
                device["last_session_date"],
                tz=timezone.utc,
            )
        return cls(
            device_type=to_enum(DeviceType, device["type"], DeviceType.UNKNOWN),
            battery=DeviceBattery(device["battery"]),
            model=to_enum(DeviceModel, device["model_id"], DeviceModel.UNKNOWN),
            first_session_date=first_session_date,
            last_session_date=last_session_date,
            device_id=device["deviceid"],
            hashed_device_id=device["hash_deviceid"],
        )


@dataclass(slots=True)
class Goals:
    """Model for Withings goals."""

    steps: int
    sleep: int
    weight: float

    @classmethod
    def from_api(cls, goals: dict[str, Any]) -> Self:
        """Initialize from the API."""
        return cls(
            steps=goals["steps"],
            sleep=goals["sleep"],
            weight=get_measurement_from_dict(goals["weight"]),
        )


class MeasurementGroupAttribution(IntEnum):
    """Measure group attributions."""

    UNKNOWN = -1
    DEVICE_ENTRY_FOR_USER = 0
    DEVICE_ENTRY_FOR_USER_AMBIGUOUS = 1
    MANUAL_USER_ENTRY = 2
    MANUAL_USER_DURING_ACCOUNT_CREATION = 4
    MEASURE_AUTO = 5
    MEASURE_USER_CONFIRMED = 7


class MeasurementGroupCategory(IntEnum):
    """Measure categories."""

    UNKNOWN = 0
    REAL = 1
    USER_OBJECTIVES = 2


@dataclass(slots=True)
class MeasurementGroup:
    """Model for a measurement group."""

    group_id: int
    attribution: MeasurementGroupAttribution
    taken_at: datetime
    stored_at: datetime
    updated_at: datetime
    category: MeasurementGroupCategory
    device_id: str
    hashed_device_id: str
    measurements: list[Measurement]

    @classmethod
    def from_api(cls, measurement_group: dict[str, Any]) -> Self:
        """Initialize from the API."""
        if measurement_group["attrib"] == 8:
            measurement_group["attrib"] = 0
        return cls(
            group_id=measurement_group["grpid"],
            attribution=to_enum(
                MeasurementGroupAttribution,
                measurement_group["attrib"],
                MeasurementGroupAttribution.UNKNOWN,
            ),
            taken_at=datetime.fromtimestamp(
                measurement_group["date"],
                tz=timezone.utc,
            ),
            stored_at=datetime.fromtimestamp(
                measurement_group["created"],
                tz=timezone.utc,
            ),
            updated_at=datetime.fromtimestamp(
                measurement_group["modified"],
                tz=timezone.utc,
            ),
            category=to_enum(
                MeasurementGroupCategory,
                measurement_group["category"],
                MeasurementGroupCategory.UNKNOWN,
            ),
            device_id=measurement_group["deviceid"],
            hashed_device_id=measurement_group["hash_deviceid"],
            measurements=[
                Measurement.from_api(measurement)
                for measurement in measurement_group["measures"]
            ],
        )


class MeasurementType(IntEnum):
    """Measurement types."""

    UNKNOWN = 0
    WEIGHT = 1
    HEIGHT = 4
    FAT_FREE_MASS = 5
    FAT_RATIO = 6
    FAT_MASS_WEIGHT = 8
    DIASTOLIC_BLOOD_PRESSURE = 9
    SYSTOLIC_BLOOD_PRESSURE = 10
    HEART_RATE = 11
    TEMPERATURE = 12
    SP02 = 54
    BODY_TEMPERATURE = 71
    SKIN_TEMPERATURE = 73
    MUSCLE_MASS = 76
    HYDRATION = 77
    BONE_MASS = 88
    PULSE_WAVE_VELOCITY = 91
    VO2 = 123
    ATRIAL_FIBRILLATION = 130
    QRS_INTERVAL = 135
    PR_INTERVAL = 136
    QT_INTERVAL = 137
    CORRECTED_QT_INTERVAL = 138
    ATRIAL_FIBRILLATION_FROM_PPG = 139
    VASCULAR_AGE = 155
    NERVE_HEALTH_SCORE_LEFT_FOOT = 158
    NERVE_HEALTH_SCORE_RIGHT_FOOT = 159
    NERVE_HEALTH_SCORE_FEET = 167
    EXTRACELLULAR_WATER = 168
    INTRACELLULAR_WATER = 169
    VISCERAL_FAT = 170
    FAT_MASS_FOR_SEGMENTS = 174
    MUSCLE_MASS_FOR_SEGMENTS = 175
    ELECTRODERMAL_ACTIVITY_FEET = 196
    ELECTRODERMAL_ACTIVITY_LEFT_FOOT = 197
    ELECTRODERMAL_ACTIVITY_RIGHT_FOOT = 198


@dataclass(slots=True)
class Measurement:
    """Model for a measurement."""

    measurement_type: MeasurementType
    value: float

    @classmethod
    def from_api(cls, measurement: dict[str, Any]) -> Self:
        """Initialize from the API."""
        return cls(
            measurement_type=to_enum(
                MeasurementType,
                measurement["type"],
                MeasurementType.UNKNOWN,
            ),
            value=get_measurement_from_dict(measurement),
        )


class NotificationCategory(IntEnum):
    """Enum representing the notification category (Appli)."""

    UNKNOWN = 0
    WEIGHT = 1
    TEMPERATURE = 2
    PRESSURE = 4
    ACTIVITY = 16
    SLEEP = 44
    USER_DATA = 46
    IN_BED = 50
    OUT_BED = 51
    INITIAL_INFLATION_DONE = 52
    NO_ACCOUNT_ASSOCIATED = 53
    ECG = 54
    ECG_FAILED = 55
    GLUCOSE = 58


class Services(IntFlag):
    """Enum representing the possible services."""

    MEASURE_GET_MEAS = 1
    MEASURE_V2_GET_ACTIVITY = 2
    MEASURE_V2_GET_INTRA_DAY_ACTIVITY = 4
    MEASURE_V2_GET_WORKOUTS = 8
    SLEEP_V2_GET_SLEEP = 16
    SLEEP_V2_GET_SUMMARY = 32
    USER_V2_ACTIVATE = 64
    USER_V2_LINK = 128
    HEART_V2_LIST = 256


def get_measurement_type_from_notification_category(
    category: NotificationCategory,
) -> list[MeasurementType]:
    """Get measurement types from notification category."""
    match category:
        case NotificationCategory.WEIGHT:
            return [
                MeasurementType.WEIGHT,
                MeasurementType.FAT_FREE_MASS,
                MeasurementType.FAT_RATIO,
                MeasurementType.FAT_MASS_WEIGHT,
                MeasurementType.BODY_TEMPERATURE,
                MeasurementType.SKIN_TEMPERATURE,
                MeasurementType.MUSCLE_MASS,
                MeasurementType.HYDRATION,
                MeasurementType.BONE_MASS,
                MeasurementType.PULSE_WAVE_VELOCITY,
            ]
        case NotificationCategory.TEMPERATURE:
            return [
                MeasurementType.TEMPERATURE,
                MeasurementType.BODY_TEMPERATURE,
                MeasurementType.SKIN_TEMPERATURE,
            ]
        case NotificationCategory.PRESSURE:
            return [
                MeasurementType.DIASTOLIC_BLOOD_PRESSURE,
                MeasurementType.SYSTOLIC_BLOOD_PRESSURE,
                MeasurementType.HEART_RATE,
                MeasurementType.SP02,
            ]
    return []


@dataclass(slots=True)
class NotificationConfiguration:
    """Class representing the webhook config from Withings."""

    notification_category: NotificationCategory
    callback_url: str
    expires: datetime
    comment: str

    @classmethod
    def from_api(cls, configuration: dict[str, Any]) -> Self:
        """Initialize from the API."""
        return cls(
            notification_category=to_enum(
                NotificationCategory,
                configuration["appli"],
                NotificationCategory.UNKNOWN,
            ),
            callback_url=configuration["callbackurl"],
            expires=datetime.fromtimestamp(
                configuration["expires"],
                tz=timezone.utc,
            ),
            comment=configuration["comment"],
        )


@dataclass(slots=True)
class WebhookCall:
    """Class representing the webhook call from Withings."""

    user_id: int
    notification_category: NotificationCategory
    start_date: datetime
    end_date: datetime

    @classmethod
    def from_api(cls, data: dict[str, Any]) -> Self:
        """Initialize from the API."""
        return cls(
            user_id=data["userid"],
            notification_category=to_enum(
                NotificationCategory,
                data["appli"],
                NotificationCategory.UNKNOWN,
            ),
            start_date=datetime.fromtimestamp(
                data["startdate"],
                tz=timezone.utc,
            ),
            end_date=datetime.fromtimestamp(
                data["enddate"],
                tz=timezone.utc,
            ),
        )
