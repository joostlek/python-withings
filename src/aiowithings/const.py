"""Asynchronous Python client for Withings."""
from __future__ import annotations

import logging

LOGGER = logging.getLogger(__package__)

STATUS_SUCCESS: set[int] = {0}

STATUS_AUTH_FAILED: set[int] = {100, 101, 102, 200, 401}

STATUS_INVALID_PARAMS: set[int] = {
    201,
    202,
    203,
    204,
    205,
    206,
    207,
    208,
    209,
    210,
    211,
    212,
    213,
    216,
    217,
    218,
    220,
    221,
    223,
    225,
    227,
    228,
    229,
    230,
    234,
    235,
    236,
    238,
    240,
    241,
    242,
    243,
    244,
    245,
    246,
    247,
    248,
    249,
    250,
    251,
    252,
    254,
    260,
    261,
    262,
    263,
    264,
    265,
    266,
    267,
    271,
    272,
    275,
    276,
    283,
    284,
    285,
    286,
    287,
    288,
    290,
    293,
    294,
    295,
    297,
    300,
    301,
    302,
    303,
    304,
    321,
    323,
    324,
    325,
    326,
    327,
    328,
    329,
    330,
    331,
    332,
    333,
    334,
    335,
    336,
    337,
    338,
    339,
    340,
    341,
    342,
    343,
    344,
    345,
    346,
    347,
    348,
    349,
    350,
    351,
    352,
    353,
    380,
    381,
    382,
    400,
    501,
    502,
    503,
    504,
    505,
    506,
    509,
    510,
    511,
    523,
    532,
    3017,
    3018,
    3019,
}

STATUS_UNAUTHORIZED: set[int] = {214, 277, 2553, 2554, 2555}

STATUS_ERROR_OCCURRED: set[int] = {
    215,
    219,
    222,
    224,
    226,
    231,
    233,
    237,
    253,
    255,
    256,
    257,
    258,
    259,
    268,
    269,
    270,
    273,
    274,
    278,
    279,
    280,
    281,
    282,
    289,
    291,
    292,
    296,
    298,
    305,
    306,
    308,
    309,
    310,
    311,
    312,
    313,
    314,
    315,
    316,
    317,
    318,
    319,
    320,
    322,
    370,
    371,
    372,
    373,
    374,
    375,
    383,
    391,
    402,
    516,
    517,
    518,
    519,
    520,
    521,
    525,
    526,
    527,
    528,
    529,
    530,
    531,
    533,
    602,
    700,
    1051,
    1052,
    1053,
    1054,
    2551,
    2552,
    2556,
    2557,
    2558,
    2559,
    3000,
    3001,
    3002,
    3003,
    3004,
    3005,
    3006,
    3007,
    3008,
    3009,
    3010,
    3011,
    3012,
    3013,
    3014,
    3015,
    3016,
    3020,
    3021,
    3022,
    3023,
    3024,
    5000,
    5001,
    5005,
    5006,
    6000,
    6010,
    6011,
    9000,
    10000,
}

STATUS_TIMEOUT: set[int] = {522}
STATUS_BAD_STATE: set[int] = {524}
STATUS_TOO_MANY_REQUESTS: set[int] = {601}
