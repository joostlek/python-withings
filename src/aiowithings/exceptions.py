"""Asynchronous Python client for Withings."""


class WithingsError(Exception):
    """Generic exception."""


class WithingsConnectionError(WithingsError):
    """Withings connection exception."""


class WithingsAuthenticationFailedError(WithingsError):
    """Withings authentication failed exception."""


class WithingsInvalidParamsError(WithingsError):
    """Withings invalid params exception."""


class WithingsUnauthorizedError(WithingsError):
    """Withings unauthorized exception."""


class WithingsErrorOccurredError(WithingsError):
    """Withings error occurred exception."""


class WithingsBadStateError(WithingsError):
    """Withings bad state exception."""


class WithingsTooManyRequestsError(WithingsError):
    """Withings too many requests exception."""


class WithingsUnknownStatusError(WithingsError):
    """Withings unknown status exception."""
