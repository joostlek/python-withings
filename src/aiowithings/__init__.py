"""Asynchronous Python client for Withings."""
from .const import AUTHORIZATION_URL, TOKEN_URL
from .exceptions import (
    WithingsAuthenticationFailedError,
    WithingsBadStateError,
    WithingsConnectionError,
    WithingsError,
    WithingsErrorOccurredError,
    WithingsInvalidParamsError,
    WithingsTooManyRequestsError,
    WithingsUnauthorizedError,
    WithingsUnknownStatusError,
)
from .helpers import aggregate_measurements
from .models import (
    Activity,
    ActivityDataFields,
    ActivityDataOrigin,
    AuthScope,
    Device,
    DeviceBattery,
    DeviceModel,
    DeviceType,
    Goals,
    Measurement,
    MeasurementAttribution,
    MeasurementGroup,
    MeasurementGroupCategory,
    MeasurementType,
    NotificationCategory,
    Services,
    SleepDataFields,
    SleepSeries,
    SleepSeriesTimeData,
    SleepState,
    SleepSummary,
    SleepSummaryDataFields,
    WebhookCall,
    Workout,
    WorkoutCategory,
    WorkoutDataFields,
    get_measurement_type_from_notification_category,
)
from .withings import WithingsClient

__all__ = [
    "AUTHORIZATION_URL",
    "TOKEN_URL",
    "AuthScope",
    "ActivityDataFields",
    "ActivityDataOrigin",
    "Activity",
    "DeviceBattery",
    "Device",
    "DeviceType",
    "DeviceModel",
    "WithingsError",
    "WithingsConnectionError",
    "WithingsAuthenticationFailedError",
    "WithingsInvalidParamsError",
    "WithingsUnauthorizedError",
    "WithingsErrorOccurredError",
    "WithingsBadStateError",
    "WithingsTooManyRequestsError",
    "WithingsUnknownStatusError",
    "WithingsClient",
    "Goals",
    "MeasurementAttribution",
    "MeasurementGroupCategory",
    "MeasurementGroup",
    "MeasurementType",
    "Measurement",
    "get_measurement_type_from_notification_category",
    "aggregate_measurements",
    "NotificationCategory",
    "Services",
    "WebhookCall",
    "SleepSummary",
    "SleepSummaryDataFields",
    "SleepState",
    "SleepSeriesTimeData",
    "SleepSeries",
    "SleepDataFields",
    "WorkoutDataFields",
    "Workout",
    "WorkoutCategory",
]
