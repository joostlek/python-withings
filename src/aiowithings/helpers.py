"""Asynchronous Python client for Withings."""
from __future__ import annotations

from typing import TYPE_CHECKING

from aiowithings.models import MeasurementAttribution, MeasurementType, SleepSummary

if TYPE_CHECKING:
    from aiowithings import MeasurementGroup


def aggregate_measurements(
    measurements: list[MeasurementGroup],
) -> dict[MeasurementType, float]:
    """Aggregate the measurements to return a list of the latest measurements."""
    result: dict[MeasurementType, float] = {}

    groups = sorted(measurements, key=lambda group: group.taken_at)

    for measurement in groups:
        if measurement.attribution not in (
            MeasurementAttribution.UNKNOWN,
            MeasurementAttribution.DEVICE_ENTRY_FOR_USER_AMBIGUOUS,
        ):
            for data_point in measurement.measurements:
                result[data_point.measurement_type] = data_point.value

    return result


def aggregate_sleep_summary(
    sleep_summaries: list[SleepSummary],
) -> SleepSummary | None:
    """Aggregate the sleep summaries to return a combined report."""
    if len(sleep_summaries) == 0:
        return None
    if len(sleep_summaries) == 1:
        return sleep_summaries[0]
    sorted_sleep_summaries = sorted(
        sleep_summaries,
        key=lambda sleep_summary: sleep_summary.start_date,
    )

    def average(data: list[int]) -> int:
        return round(sum(data) / len(data))

    return SleepSummary(
        start_date=sorted_sleep_summaries[0].start_date,
        end_date=sorted_sleep_summaries[1].end_date,
        date=sorted_sleep_summaries[0].date,
        hashed_device_id=sorted_sleep_summaries[0].hashed_device_id,
        breathing_disturbances_intensity=average(
            [
                summary.breathing_disturbances_intensity
                for summary in sleep_summaries
                if summary.breathing_disturbances_intensity is not None
            ],
        ),
        deep_sleep_duration=sum(
            summary.deep_sleep_duration
            for summary in sleep_summaries
            if summary.deep_sleep_duration is not None
        ),
        sleep_latency=average(
            [
                summary.sleep_latency
                for summary in sleep_summaries
                if summary.sleep_latency is not None
            ],
        ),
        wake_up_latency=average(
            [
                summary.wake_up_latency
                for summary in sleep_summaries
                if summary.wake_up_latency is not None
            ],
        ),
        average_heart_rate=average(
            [
                summary.average_heart_rate
                for summary in sleep_summaries
                if summary.average_heart_rate is not None
            ],
        ),
        max_heart_rate=average(
            [
                summary.max_heart_rate
                for summary in sleep_summaries
                if summary.max_heart_rate is not None
            ],
        ),
        min_heart_rate=average(
            [
                summary.min_heart_rate
                for summary in sleep_summaries
                if summary.min_heart_rate is not None
            ],
        ),
        light_sleep_duration=sum(
            summary.light_sleep_duration
            for summary in sleep_summaries
            if summary.light_sleep_duration is not None
        ),
        rem_sleep_duration=sum(
            summary.rem_sleep_duration
            for summary in sleep_summaries
            if summary.rem_sleep_duration is not None
        ),
        average_respiration_rate=average(
            [
                summary.average_respiration_rate
                for summary in sleep_summaries
                if summary.average_respiration_rate is not None
            ],
        ),
        max_respiration_rate=average(
            [
                summary.max_respiration_rate
                for summary in sleep_summaries
                if summary.max_respiration_rate is not None
            ],
        ),
        min_respiration_rate=average(
            [
                summary.min_respiration_rate
                for summary in sleep_summaries
                if summary.min_respiration_rate is not None
            ],
        ),
        sleep_score=max(
            (
                summary.sleep_score
                for summary in sleep_summaries
                if summary.sleep_score is not None
            ),
            default=0,
        ),
        snoring=average(
            [
                summary.snoring
                for summary in sleep_summaries
                if summary.snoring is not None
            ],
        ),
        snoring_count=sum(
            summary.snoring_count
            for summary in sleep_summaries
            if summary.snoring_count is not None
        ),
        wake_up_count=sum(
            summary.wake_up_count
            for summary in sleep_summaries
            if summary.wake_up_count is not None
        ),
        total_time_awake=average(
            [
                summary.total_time_awake
                for summary in sleep_summaries
                if summary.total_time_awake is not None
            ],
        ),
        active_movement_duration=None,
        apnea_hypopnea_index=None,
        average_movement_score=None,
        external_time_asleep=None,
        out_of_bed_count=None,
        rem_sleep_phase_count=None,
        sleep_efficiency=None,
        time_awake_during_sleep=None,
        total_sleep_time=None,
        total_time_in_bed=None,
        withings_index=None,
    )
