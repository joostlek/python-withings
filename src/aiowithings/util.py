"""Asynchronous Python client for Withings."""
from __future__ import annotations

from typing import Any, TypeVar, cast

from aiowithings.const import LOGGER

_EnumT = TypeVar("_EnumT")


def to_enum(
    enum_class: type[_EnumT],
    value: Any,
    default_value: _EnumT,
) -> _EnumT:
    """Convert a value to an enum and log if it doesn't exist."""
    try:
        return enum_class(value)  # type: ignore[call-arg]
    except ValueError:
        LOGGER.warning(
            "%s is an unsupported value for %s, please report this at https://github.com/joostlek/python-withings/issues",
            value,
            str(enum_class),
        )
        return default_value


def get_measurement(value: int, unit: int) -> float:
    """Convert value and unit to real value."""
    return cast(float, value * pow(10, unit))


def get_measurement_from_dict(input_dict: dict[str, Any]) -> float:
    """Convert dict with value and unit to real value."""
    value = cast(int, input_dict["value"])
    unit = cast(int, input_dict["unit"])
    return get_measurement(value, unit)
