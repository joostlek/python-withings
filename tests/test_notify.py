"""Asynchronous Python client for Withings."""

import aiohttp
from aresponses import ResponsesMockServer
import pytest

from aiowithings import (
    NotificationCategory,
    WebhookCall,
    WithingsClient,
    get_measurement_type_from_notification_category,
)
from syrupy import SnapshotAssertion

from . import load_fixture

WITHINGS_URL = "wbsapi.withings.net"


async def test_subscribing(
    aresponses: ResponsesMockServer,
) -> None:
    """Test subscribing to webhook updates."""
    aresponses.add(
        WITHINGS_URL,
        "/notify",
        "POST",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("notify_subscribe.json"),
        ),
        body_pattern="action=subscribe&callbackurl=https%3A%2F%2Ftest.com%2Fcallback&appli=4",
    )
    async with aiohttp.ClientSession() as session:
        withings = WithingsClient(session=session)
        withings.authenticate("test")
        await withings.subscribe_notification(
            "https://test.com/callback",
            NotificationCategory.PRESSURE,
        )
        await withings.close()


async def test_revoking(
    aresponses: ResponsesMockServer,
) -> None:
    """Test subscribing to webhook updates."""
    aresponses.add(
        WITHINGS_URL,
        "/notify",
        "POST",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("notify_revoke.json"),
        ),
        body_pattern="action=revoke&callbackurl=https%3A%2F%2Ftest.com%2Fcallback&appli=4",
    )
    async with aiohttp.ClientSession() as session:
        withings = WithingsClient(session=session)
        withings.authenticate("test")
        await withings.revoke_notification_configurations(
            "https://test.com/callback",
            NotificationCategory.PRESSURE,
        )
        await withings.close()


async def test_list_subscriptions(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving subscriptions."""
    aresponses.add(
        WITHINGS_URL,
        "/notify",
        "POST",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("notify_list.json"),
        ),
        body_pattern="action=list&appli=1",
    )
    async with aiohttp.ClientSession() as session:
        withings = WithingsClient(session=session)
        withings.authenticate("test")
        response = await withings.list_notification_configurations(
            NotificationCategory.WEIGHT,
        )
        assert response == snapshot
        await withings.close()


@pytest.mark.parametrize(
    "notification_category",
    list(NotificationCategory),
    ids=[nc.name for nc in NotificationCategory],
)
async def test_measurement_points_to_get(
    notification_category: NotificationCategory,
    snapshot: SnapshotAssertion,
) -> None:
    """Test if we receive the right updated measurement points."""
    assert (
        get_measurement_type_from_notification_category(notification_category)
        == snapshot
    )


async def test_webhook_object(snapshot: SnapshotAssertion) -> None:
    """Test if the webhook call transforms in a good object."""
    data = {
        "userid": 12345,
        "appli": 1,
        "startdate": 1530576000,
        "enddate": 1530698753,
    }
    assert WebhookCall.from_api(data) == snapshot
