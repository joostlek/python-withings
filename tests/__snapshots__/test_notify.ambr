# serializer version: 1
# name: test_list_subscriptions
  list([
    NotificationConfiguration(notification_category=<NotificationCategory.WEIGHT: 1>, callback_url='https://www.withings.com', expires=datetime.datetime(2038, 1, 19, 3, 14, 7, tzinfo=datetime.timezone.utc), comment='My notification subscription'),
  ])
# ---
# name: test_measurement_points_to_get[ACTIVITY]
  list([
  ])
# ---
# name: test_measurement_points_to_get[ECG]
  list([
  ])
# ---
# name: test_measurement_points_to_get[ECG_FAILED]
  list([
  ])
# ---
# name: test_measurement_points_to_get[GLUCOSE]
  list([
  ])
# ---
# name: test_measurement_points_to_get[INITIAL_INFLATION_DONE]
  list([
  ])
# ---
# name: test_measurement_points_to_get[IN_BED]
  list([
  ])
# ---
# name: test_measurement_points_to_get[NO_ACCOUNT_ASSOCIATED]
  list([
  ])
# ---
# name: test_measurement_points_to_get[OUT_BED]
  list([
  ])
# ---
# name: test_measurement_points_to_get[PRESSURE]
  list([
    <MeasurementType.DIASTOLIC_BLOOD_PRESSURE: 9>,
    <MeasurementType.SYSTOLIC_BLOOD_PRESSURE: 10>,
    <MeasurementType.HEART_RATE: 11>,
    <MeasurementType.SP02: 54>,
  ])
# ---
# name: test_measurement_points_to_get[SLEEP]
  list([
  ])
# ---
# name: test_measurement_points_to_get[TEMPERATURE]
  list([
    <MeasurementType.TEMPERATURE: 12>,
    <MeasurementType.BODY_TEMPERATURE: 71>,
    <MeasurementType.SKIN_TEMPERATURE: 73>,
  ])
# ---
# name: test_measurement_points_to_get[UNKNOWN]
  list([
  ])
# ---
# name: test_measurement_points_to_get[USER_DATA]
  list([
  ])
# ---
# name: test_measurement_points_to_get[WEIGHT]
  list([
    <MeasurementType.WEIGHT: 1>,
    <MeasurementType.FAT_FREE_MASS: 5>,
    <MeasurementType.FAT_RATIO: 6>,
    <MeasurementType.FAT_MASS_WEIGHT: 8>,
    <MeasurementType.BODY_TEMPERATURE: 71>,
    <MeasurementType.SKIN_TEMPERATURE: 73>,
    <MeasurementType.MUSCLE_MASS: 76>,
    <MeasurementType.HYDRATION: 77>,
    <MeasurementType.BONE_MASS: 88>,
    <MeasurementType.PULSE_WAVE_VELOCITY: 91>,
  ])
# ---
# name: test_webhook_object
  WebhookCall(user_id=12345, notification_category=<NotificationCategory.WEIGHT: 1>, start_date=datetime.datetime(2018, 7, 3, 0, 0, tzinfo=datetime.timezone.utc), end_date=datetime.datetime(2018, 7, 4, 10, 5, 53, tzinfo=datetime.timezone.utc))
# ---
